spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/ApplicationManifestDifferenceFactory.java

```diff
@@ -0,0 +1,132 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.skipper.server.deployer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import com.google.common.collect.MapDifference;
+import com.google.common.collect.Maps;
+
+import org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationManifest;
+import org.springframework.cloud.skipper.domain.deployer.ApplicationManifestDifference;
+
+import org.springframework.util.Assert;
+
+/**
+ * A Factory to create an {@link ApplicationManifestDifference}.
+ * @author Mark Pollack
+ */
+public class ApplicationManifestDifferenceFactory {
+
+	/**
+	 * Create an ApplicationManifestDifference for an application given two application
+	 * manifests
+	 * @param applicationName the name of the application represented in each manifest
+	 * @param existingApplicationManifest the manifest of the currently deployed application
+	 * @param replacingApplicationManifest the manifest of the application that is a candidate
+	 * to replace the currently deployed application
+	 * @return The ApplicationManifestDifference for the application
+	 */
+	public ApplicationManifestDifference createApplicationManifestDifference(
+			String applicationName,
+			SpringCloudDeployerApplicationManifest existingApplicationManifest,
+			SpringCloudDeployerApplicationManifest replacingApplicationManifest) {
+
+		Assert.notNull(applicationName, "applicationName can not be null");
+		Assert.notNull(existingApplicationManifest, "existingApplicationManifest can not be null");
+		Assert.notNull(replacingApplicationManifest, "replacingApplicationManifest can not be null");
+
+		return new ApplicationManifestDifference(applicationName,
+				getApiAndKindDifference(existingApplicationManifest, replacingApplicationManifest),
+				getMetadataDifference(existingApplicationManifest, replacingApplicationManifest),
+				getResourceAndVersionDifference(existingApplicationManifest, replacingApplicationManifest),
+				getApplicationPropertiesDifference(existingApplicationManifest, replacingApplicationManifest),
+				getDeploymentPropertiesDifference(existingApplicationManifest, replacingApplicationManifest));
+	}
+
+	protected MapDifference getApiAndKindDifference(
+			SpringCloudDeployerApplicationManifest existingApplicationManifest,
+			SpringCloudDeployerApplicationManifest replacingApplicationManifest) {
+		Map<String, String> existingApiAndKindMap = new HashMap<>();
+		existingApiAndKindMap.put("api", existingApplicationManifest.getApiVersion().trim());
```

Comment:
Isn't this `apiVersion` instead of `api`? I think we need to use the constants defined in `SpringCloudDeployerApplicationManifest` instead of String values for the keys.

Reply:
ok, I see that as long as we use the same key `api` in the other comparing (replacingApplicationManifest) map we are good. But, we can still use `apiVersion` for consistency I guess.

Reply:
Sure, will change to `apiVersion`